src/shogun/optimization/StanFirstOrderSAGCostFunction.cpp

```diff
@@ -0,0 +1,140 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Authors: Elfarouk
+ */
+
+#include <shogun/optimization/StanFirstOrderSAGCostFunction.h>
+#include <shogun/base/range.h>
+#include <shogun/mathematics/Math.h>
+using namespace shogun;
+using stan::math::var;
+using std::function;
+using Eigen::Matrix;
+using Eigen::Dynamic;
+
+StanFirstOrderSAGCostFunction::StanFirstOrderSAGCostFunction(
+    SGMatrix<float64_t> X, SGMatrix<float64_t> y,
+    StanVector* trainable_parameters,
+    StanFunctionsVector<float64_t>* cost_for_ith_point,
+    FunctionReturnsStan<StanVector*>* total_cost)
+{
+	REQUIRE(X.size() > 0, "Empty X provided");
+	REQUIRE(y.size() > 0, "Empty y provided");
+	auto num_of_variables = trainable_parameters->rows();
+	REQUIRE(
+	    num_of_variables > 0, "Provided %d variables in the parameters, more "
+	                          "than 0 parameters required",
+	    num_of_variables);
+	REQUIRE(cost_for_ith_point != NULL, "Cost for ith point is not provided");
+	REQUIRE(total_cost != NULL, "Total cost function is not provided");
+	m_X = X;
+	m_y = y;
+	m_trainable_parameters = trainable_parameters;
+	m_cost_for_ith_point = cost_for_ith_point;
+	m_total_cost = total_cost;
+	m_ref_trainable_parameters = SGVector<float64_t>(num_of_variables);
+	for (auto i : range(num_of_variables))
+	{
+		m_ref_trainable_parameters[i] = (*m_trainable_parameters)(i, 0).val();
+	}
+}
+
+void StanFirstOrderSAGCostFunction::set_training_data(
+    SGMatrix<float64_t> X_new, SGMatrix<float64_t> y_new)
+{
+	REQUIRE(X_new.size() > 0, "Empty X provided");
+	REQUIRE(y_new.size() > 0, "Empty y provided");
+	this->m_X = X_new;
+	this->m_y = y_new;
+}
+
+StanFirstOrderSAGCostFunction::~StanFirstOrderSAGCostFunction()
+{
+}
+
+void StanFirstOrderSAGCostFunction::begin_sample()
+{
+	m_index_of_sample = -1;
+}
+
+bool StanFirstOrderSAGCostFunction::next_sample()
+{
+	++m_index_of_sample;
+	return m_index_of_sample < get_sample_size();
+}
+
+void StanFirstOrderSAGCostFunction::update_stan_vectors_to_reference_values()
+{
+	auto num_of_variables = m_trainable_parameters->rows();
+	for (auto i : range(num_of_variables))
+	{
+		(*m_trainable_parameters)(i, 0) = m_ref_trainable_parameters[i];
+	}
+}
+SGVector<float64_t> StanFirstOrderSAGCostFunction::get_gradient()
+{
+	auto num_of_variables = m_trainable_parameters->rows();
+	REQUIRE(
+	    num_of_variables > 0,
+	    "Number of sample must be greater than 0, you provided no samples");
+
+	update_stan_vectors_to_reference_values();
+	var f_i = (*m_cost_for_ith_point)(m_index_of_sample, 0)(
+	    m_trainable_parameters, m_index_of_sample);
+
+	stan::math::set_zero_all_adjoints();
+	f_i.grad();
+
+	SGVector<float64_t>::EigenVectorXt gradients =
+	    m_trainable_parameters->unaryExpr(
+	        [](stan::math::var x) -> float64_t { return x.adj(); });
+	// clone needed because gradients is local variable
+	return SGVector<float64_t>(gradients).clone();
```

Comment:
you could just simply wrap the EigenVectorXt with SGVector and not do cloning...
namely
```
SGVector<float64>(gradients.data(), gradients.size())
```
but i reckon it'd be better that simply just create an

```
SGVector<float64_t> gradients(size);
SGVector<float64_t>::EigenVectorXt mapped_gradients = gradients;
/// now you can use mapped_gradients as an eigen vector

return gradients;
```

Reply:
@vigsterkr  But the data was being destroyed since the variable was local. I tried the second one but when I returned gradients, it just contained garbage values?

Reply:
if you return SGVector<float64_t> that should trigger a copy-ctor that should ++ the ref counter and hence shouldn't delete the data itself.

Reply:
there should be many examples where this is used in shogun.... e.g. https://github.com/shogun-toolbox/shogun/blob/develop/src/shogun/machine/gp/GaussianLikelihood.cpp#L238